Data race in `ParallelReadBuffer`/`copyFromIStreamWithProgressCallback`

https://s3.amazonaws.com/clickhouse-test-reports/51268/be3352a078908a907bb2f3eb57c3711a35e7178c/integration_tests__tsan__[5_6].html


```
E           Exception: Sanitizer assert found for instance ==================
E           WARNING: ThreadSanitizer: data race (pid=1)
E             Write of size 1 at 0x7f623c0fd000 by thread T281:
E               #0 std::__1::basic_streambuf<char, std::__1::char_traits<char>>::xsgetn(char*, long) build_docker/./contrib/llvm-project/libcxx/include/streambuf:424:18 (clickhouse+0x28deb737) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #1 std::__1::basic_streambuf<char, std::__1::char_traits<char>>::sgetn[abi:v15000](char*, long) build_docker/./contrib/llvm-project/libcxx/include/streambuf:204:14 (clickhouse+0x28deefd0) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #2 std::__1::basic_istream<char, std::__1::char_traits<char>>::read(char*, long) build_docker/./contrib/llvm-project/libcxx/include/istream:1052:36 (clickhouse+0x28deefd0)
E               #3 DB::copyFromIStreamWithProgressCallback(std::__1::basic_istream<char, std::__1::char_traits<char>>&, char*, unsigned long, std::__1::function<bool (unsigned long)> const&, bool*) build_docker/./src/IO/SeekableReadBuffer.cpp:77:14 (clickhouse+0x1a411d81) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #4 DB::detail::ReadWriteBufferFromHTTPBase<std::__1::shared_ptr<DB::UpdatableSession<DB::SessionFactory>>>::readBigAt(char*, unsigned long, unsigned long, std::__1::function<bool (unsigned long)> const&) build_docker/./src/IO/ReadWriteBufferFromHTTP.cpp:636:24 (clickhouse+0x1a40420f) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #5 DB::ParallelReadBuffer::readerThreadFunction(std::__1::shared_ptr<DB::ParallelReadBuffer::ReadWorker>) build_docker/./src/IO/ParallelReadBuffer.cpp:259:26 (clickhouse+0x160a3875) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #6 DB::ParallelReadBuffer::addReaderToPool()::$_0::operator()() build_docker/./src/IO/ParallelReadBuffer.cpp:77:64 (clickhouse+0x160a3ee6) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #7 decltype(std::declval<DB::ParallelReadBuffer::addReaderToPool()::$_0&>()()) std::__1::__invoke[abi:v15000]<DB::ParallelReadBuffer::addReaderToPool()::$_0&>(DB::ParallelReadBuffer::addReaderToPool()::$_0&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x160a3ee6)
E               #8 void std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::ParallelReadBuffer::addReaderToPool()::$_0&>(DB::ParallelReadBuffer::addReaderToPool()::$_0&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9 (clickhouse+0x160a3ee6)
E               #9 std::__1::__function::__default_alloc_func<DB::ParallelReadBuffer::addReaderToPool()::$_0, void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12 (clickhouse+0x160a3ee6)
E               #10 void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::ParallelReadBuffer::addReaderToPool()::$_0, void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16 (clickhouse+0x160a3ee6)
E               #11 std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16 (clickhouse+0x1e0f93bd) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #12 std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12 (clickhouse+0x1e0f93bd)
E               #13 std::__1::function<std::__1::future<void> (std::__1::function<void ()>&&, Priority)> DB::threadPoolCallbackRunner<void, std::__1::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(std::__1::function<void ()>&&, Priority)::operator()(std::__1::function<void ()>&&, Priority)::'lambda'()::operator()() build_docker/./src/Interpreters/threadPoolCallbackRunner.h:42:20 (clickhouse+0x1e0f93bd)
E               #14 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<std::__1::function<std::__1::future<void> (std::__1::function<void ()>&&, Priority)> DB::threadPoolCallbackRunner<void, std::__1::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(std::__1::function<void ()>&&, Priority)::operator()(std::__1::function<void ()>&&, Priority)::'lambda'()&>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x1e0f93bd)
E               #15 std::__1::__packaged_task_func<std::__1::function<std::__1::future<void> (std::__1::function<void ()>&&, Priority)> DB::threadPoolCallbackRunner<void, std::__1::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(std::__1::function<void ()>&&, Priority)::operator()(std::__1::function<void ()>&&, Priority)::'lambda'(), std::__1::allocator<std::__1::function<std::__1::future<void> (std::__1::function<void ()>&&, Priority)> DB::threadPoolCallbackRunner<void, std::__1::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(std::__1::function<void ()>&&, Priority)::operator()(std::__1::function<void ()>&&, Priority)::'lambda'()>, void ()>::operator()() build_docker/./contrib/llvm-project/libcxx/include/future:1694:12 (clickhouse+0x1e0f93bd)
E               #16 std::__1::__packaged_task_function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/future:1876:12 (clickhouse+0x1dae6fb4) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #17 std::__1::packaged_task<void ()>::operator()() build_docker/./contrib/llvm-project/libcxx/include/future:2068:9 (clickhouse+0x1dae6fb4)
E               #18 std::__1::function<std::__1::future<void> (std::__1::function<void ()>&&, Priority)> DB::threadPoolCallbackRunner<void, std::__1::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(std::__1::function<void ()>&&, Priority)::operator()(std::__1::function<void ()>&&, Priority)::'lambda0'()::operator()() const build_docker/./src/Interpreters/threadPoolCallbackRunner.h:50:70 (clickhouse+0x1e0f988c) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #19 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<std::__1::function<std::__1::future<void> (std::__1::function<void ()>&&, Priority)> DB::threadPoolCallbackRunner<void, std::__1::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(std::__1::function<void ()>&&, Priority)::operator()(std::__1::function<void ()>&&, Priority)::'lambda0'()&>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x1e0f988c)
E               #20 void std::__1::__invoke_void_return_wrapper<void, true>::__call<std::__1::function<std::__1::future<void> (std::__1::function<void ()>&&, Priority)> DB::threadPoolCallbackRunner<void, std::__1::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(std::__1::function<void ()>&&, Priority)::operator()(std::__1::function<void ()>&&, Priority)::'lambda0'()&>(std::__1::function<std::__1::future<void> (std::__1::function<void ()>&&, Priority)> DB::threadPoolCallbackRunner<void, std::__1::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(std::__1::function<void ()>&&, Priority)::operator()(std::__1::function<void ()>&&, Priority)::'lambda0'()&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9 (clickhouse+0x1e0f988c)
E               #21 std::__1::__function::__default_alloc_func<std::__1::function<std::__1::future<void> (std::__1::function<void ()>&&, Priority)> DB::threadPoolCallbackRunner<void, std::__1::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(std::__1::function<void ()>&&, Priority)::operator()(std::__1::function<void ()>&&, Priority)::'lambda0'(), void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12 (clickhouse+0x1e0f988c)
E               #22 void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<std::__1::function<std::__1::future<void> (std::__1::function<void ()>&&, Priority)> DB::threadPoolCallbackRunner<void, std::__1::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(std::__1::function<void ()>&&, Priority)::operator()(std::__1::function<void ()>&&, Priority)::'lambda0'(), void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16 (clickhouse+0x1e0f988c)
E               #23 std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16 (clickhouse+0x16123cc0) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #24 std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12 (clickhouse+0x16123cc0)
E               #25 ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__1::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) build_docker/./src/Common/ThreadPool.cpp:416:13 (clickhouse+0x16123cc0)
E               #26 void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/./src/Common/ThreadPool.cpp:180:73 (clickhouse+0x16128757) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #27 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()&>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x16128757)
E               #28 decltype(auto) std::__1::__apply_tuple_impl[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()&, std::__1::tuple<>&>(void&&, std::__1::tuple<>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/tuple:1789:1 (clickhouse+0x16128757)
E               #29 decltype(auto) std::__1::apply[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()&, std::__1::tuple<>&>(void&&, std::__1::tuple<>&) build_docker/./contrib/llvm-project/libcxx/include/tuple:1798:1 (clickhouse+0x16128757)
E               #30 ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()::operator()() build_docker/./src/Common/ThreadPool.h:229:13 (clickhouse+0x16128757)
E               #31 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x161286c1) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #32 void std::__1::__invoke_void_return_wrapper<void, true>::__call<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&>(ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9 (clickhouse+0x161286c1)
E               #33 std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12 (clickhouse+0x161286c1)
E               #34 void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16 (clickhouse+0x161286c1)
E               #35 std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16 (clickhouse+0x1611ff42) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #36 std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12 (clickhouse+0x1611ff42)
E               #37 ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) build_docker/./src/Common/ThreadPool.cpp:416:13 (clickhouse+0x1611ff42)
E               #38 void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/./src/Common/ThreadPool.cpp:180:73 (clickhouse+0x16125b31) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #39 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x16125b31)
E               #40 void std::__1::__thread_execute[abi:v15000]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(std::__1::tuple<void, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/thread:284:5 (clickhouse+0x16125b31)
E               #41 void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*) build_docker/./contrib/llvm-project/libcxx/include/thread:295:5 (clickhouse+0x16125b31)
E           
E             Previous read of size 8 at 0x7f623c0fd000 by thread T285:
E               #0 __tsan_memcpy <null> (clickhouse+0xdc58ae0) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #1 DB::saveUpToPosition(DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, char*) build_docker/./src/IO/ReadHelpers.cpp:1488:5 (clickhouse+0x1607da40) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #2 DB::loadAtPosition(DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, char*&) build_docker/./src/IO/ReadHelpers.cpp:1499:5 (clickhouse+0x1607da40)
E               #3 DB::fileSegmentationEngineTabSeparatedImpl(DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, bool, unsigned long, unsigned long, unsigned long) build_docker/./src/Processors/Formats/Impl/TabSeparatedRowInputFormat.cpp:394:12 (clickhouse+0x214fcace) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #4 DB::registerFileSegmentationEngineTabSeparated(DB::FormatFactory&)::$_0::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, bool) const::'lambda'(DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long, unsigned long)::operator()(DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long, unsigned long) const build_docker/./src/Processors/Formats/Impl/TabSeparatedRowInputFormat.cpp:447:24 (clickhouse+0x214fc629) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #5 decltype(std::declval<DB::registerFileSegmentationEngineTabSeparated(DB::FormatFactory&)::$_0::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, bool) const::'lambda'(DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long, unsigned long)&>()(std::declval<DB::ReadBuffer&>(), std::declval<DB::Memory<Allocator<false, false>>&>(), std::declval<unsigned long>(), std::declval<unsigned long>())) std::__1::__invoke[abi:v15000]<DB::registerFileSegmentationEngineTabSeparated(DB::FormatFactory&)::$_0::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, bool) const::'lambda'(DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long, unsigned long)&, DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long, unsigned long>(DB::registerFileSegmentationEngineTabSeparated(DB::FormatFactory&)::$_0::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, bool) const::'lambda'(DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long, unsigned long)&, DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long&&, unsigned long&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x214fc629)
E               #6 std::__1::pair<bool, unsigned long> std::__1::__invoke_void_return_wrapper<std::__1::pair<bool, unsigned long>, false>::__call<DB::registerFileSegmentationEngineTabSeparated(DB::FormatFactory&)::$_0::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, bool) const::'lambda'(DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long, unsigned long)&, DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long, unsigned long>(DB::registerFileSegmentationEngineTabSeparated(DB::FormatFactory&)::$_0::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, bool) const::'lambda'(DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long, unsigned long)&, DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long&&, unsigned long&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:470:16 (clickhouse+0x214fc629)
E               #7 std::__1::__function::__default_alloc_func<DB::registerFileSegmentationEngineTabSeparated(DB::FormatFactory&)::$_0::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, bool) const::'lambda'(DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long, unsigned long), std::__1::pair<bool, unsigned long> (DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long, unsigned long)>::operator()[abi:v15000](DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long&&, unsigned long&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12 (clickhouse+0x214fc629)
E               #8 std::__1::pair<bool, unsigned long> std::__1::__function::__policy_invoker<std::__1::pair<bool, unsigned long> (DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long, unsigned long)>::__call_impl<std::__1::__function::__default_alloc_func<DB::registerFileSegmentationEngineTabSeparated(DB::FormatFactory&)::$_0::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, bool) const::'lambda'(DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long, unsigned long), std::__1::pair<bool, unsigned long> (DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long, unsigned long)>>(std::__1::__function::__policy_storage const*, DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long, unsigned long) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16 (clickhouse+0x214fc629)
E               #9 std::__1::__function::__policy_func<std::__1::pair<bool, unsigned long> (DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long, unsigned long)>::operator()[abi:v15000](DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long&&, unsigned long&&) const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16 (clickhouse+0x21408649) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #10 std::__1::function<std::__1::pair<bool, unsigned long> (DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long, unsigned long)>::operator()(DB::ReadBuffer&, DB::Memory<Allocator<false, false>>&, unsigned long, unsigned long) const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12 (clickhouse+0x21408649)
E               #11 DB::ParallelParsingInputFormat::segmentatorThreadFunction(std::__1::shared_ptr<DB::ThreadGroup>) build_docker/./src/Processors/Formats/Impl/ParallelParsingInputFormat.cpp:43:58 (clickhouse+0x21408649)
E               #12 decltype(*std::declval<DB::ParallelParsingInputFormat*&>().*std::declval<void (DB::ParallelParsingInputFormat::*&)(std::__1::shared_ptr<DB::ThreadGroup>)>()(std::declval<std::__1::shared_ptr<DB::ThreadGroup>&>())) std::__1::__invoke[abi:v15000]<void (DB::ParallelParsingInputFormat::*&)(std::__1::shared_ptr<DB::ThreadGroup>), DB::ParallelParsingInputFormat*&, std::__1::shared_ptr<DB::ThreadGroup>&, void>(void (DB::ParallelParsingInputFormat::*&)(std::__1::shared_ptr<DB::ThreadGroup>), DB::ParallelParsingInputFormat*&, std::__1::shared_ptr<DB::ThreadGroup>&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:359:23 (clickhouse+0x2140e440) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #13 decltype(auto) std::__1::__apply_tuple_impl[abi:v15000]<void (DB::ParallelParsingInputFormat::*&)(std::__1::shared_ptr<DB::ThreadGroup>), std::__1::tuple<DB::ParallelParsingInputFormat*, std::__1::shared_ptr<DB::ThreadGroup>>&, 0ul, 1ul>(void (DB::ParallelParsingInputFormat::*&)(std::__1::shared_ptr<DB::ThreadGroup>), std::__1::tuple<DB::ParallelParsingInputFormat*, std::__1::shared_ptr<DB::ThreadGroup>>&, std::__1::__tuple_indices<0ul, 1ul>) build_docker/./contrib/llvm-project/libcxx/include/tuple:1789:1 (clickhouse+0x2140e440)
E               #14 decltype(auto) std::__1::apply[abi:v15000]<void (DB::ParallelParsingInputFormat::*&)(std::__1::shared_ptr<DB::ThreadGroup>), std::__1::tuple<DB::ParallelParsingInputFormat*, std::__1::shared_ptr<DB::ThreadGroup>>&>(void (DB::ParallelParsingInputFormat::*&)(std::__1::shared_ptr<DB::ThreadGroup>), std::__1::tuple<DB::ParallelParsingInputFormat*, std::__1::shared_ptr<DB::ThreadGroup>>&) build_docker/./contrib/llvm-project/libcxx/include/tuple:1798:1 (clickhouse+0x2140e440)
E               #15 ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<void (DB::ParallelParsingInputFormat::*)(std::__1::shared_ptr<DB::ThreadGroup>), DB::ParallelParsingInputFormat*, std::__1::shared_ptr<DB::ThreadGroup>>(void (DB::ParallelParsingInputFormat::*&&)(std::__1::shared_ptr<DB::ThreadGroup>), DB::ParallelParsingInputFormat*&&, std::__1::shared_ptr<DB::ThreadGroup>&&)::'lambda'()::operator()() build_docker/./src/Common/ThreadPool.h:229:13 (clickhouse+0x2140e440)
E               #16 decltype(std::declval<void (DB::ParallelParsingInputFormat::*)(std::__1::shared_ptr<DB::ThreadGroup>)>()(std::declval<DB::ParallelParsingInputFormat*>(), std::declval<std::__1::shared_ptr<DB::ThreadGroup>>())) std::__1::__invoke[abi:v15000]<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<void (DB::ParallelParsingInputFormat::*)(std::__1::shared_ptr<DB::ThreadGroup>), DB::ParallelParsingInputFormat*, std::__1::shared_ptr<DB::ThreadGroup>>(void (DB::ParallelParsingInputFormat::*&&)(std::__1::shared_ptr<DB::ThreadGroup>), DB::ParallelParsingInputFormat*&&, std::__1::shared_ptr<DB::ThreadGroup>&&)::'lambda'()&>(void (DB::ParallelParsingInputFormat::*&&)(std::__1::shared_ptr<DB::ThreadGroup>), DB::ParallelParsingInputFormat*&&, std::__1::shared_ptr<DB::ThreadGroup>&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x2140e321) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #17 void std::__1::__invoke_void_return_wrapper<void, true>::__call<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<void (DB::ParallelParsingInputFormat::*)(std::__1::shared_ptr<DB::ThreadGroup>), DB::ParallelParsingInputFormat*, std::__1::shared_ptr<DB::ThreadGroup>>(void (DB::ParallelParsingInputFormat::*&&)(std::__1::shared_ptr<DB::ThreadGroup>), DB::ParallelParsingInputFormat*&&, std::__1::shared_ptr<DB::ThreadGroup>&&)::'lambda'()&>(ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<void (DB::ParallelParsingInputFormat::*)(std::__1::shared_ptr<DB::ThreadGroup>), DB::ParallelParsingInputFormat*, std::__1::shared_ptr<DB::ThreadGroup>>(void (DB::ParallelParsingInputFormat::*&&)(std::__1::shared_ptr<DB::ThreadGroup>), DB::ParallelParsingInputFormat*&&, std::__1::shared_ptr<DB::ThreadGroup>&&)::'lambda'()&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9 (clickhouse+0x2140e321)
E               #18 std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<void (DB::ParallelParsingInputFormat::*)(std::__1::shared_ptr<DB::ThreadGroup>), DB::ParallelParsingInputFormat*, std::__1::shared_ptr<DB::ThreadGroup>>(void (DB::ParallelParsingInputFormat::*&&)(std::__1::shared_ptr<DB::ThreadGroup>), DB::ParallelParsingInputFormat*&&, std::__1::shared_ptr<DB::ThreadGroup>&&)::'lambda'(), void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12 (clickhouse+0x2140e321)
E               #19 void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<void (DB::ParallelParsingInputFormat::*)(std::__1::shared_ptr<DB::ThreadGroup>), DB::ParallelParsingInputFormat*, std::__1::shared_ptr<DB::ThreadGroup>>(void (DB::ParallelParsingInputFormat::*&&)(std::__1::shared_ptr<DB::ThreadGroup>), DB::ParallelParsingInputFormat*&&, std::__1::shared_ptr<DB::ThreadGroup>&&)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16 (clickhouse+0x2140e321)
E               #20 std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16 (clickhouse+0x1611ff42) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #21 std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12 (clickhouse+0x1611ff42)
E               #22 ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) build_docker/./src/Common/ThreadPool.cpp:416:13 (clickhouse+0x1611ff42)
E               #23 void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/./src/Common/ThreadPool.cpp:180:73 (clickhouse+0x16125b31) (BuildId: 273692bcfdeda417d2f36ecc967442bc5895b88d)
E               #24 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x16125b31)
E               #25 void std::__1::__thread_execute[abi:v15000]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(std::__1::tuple<void, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/thread:284:5 (clickhouse+0x16125b31)
E               #26 void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*) build_docker/./contrib/llvm-project/libcxx/include/thread:295:5 (clickhouse+0x16125b31)
```

Probably was introduced in https://github.com/ClickHouse/ClickHouse/pull/51165

--- a/src/IO/HTTPCommon.h
+++ b/src/IO/HTTPCommon.h
@@ -62,6 +62,7 @@ struct HTTPSessionReuseTag
 {
 };
 
+void markSessionForReuse(Poco::Net::HTTPSession & session);
 void markSessionForReuse(HTTPSessionPtr session);
 void markSessionForReuse(PooledHTTPSessionPtr session);
 

--- a/src/IO/ReadBufferFromS3.cpp
+++ b/src/IO/ReadBufferFromS3.cpp
@@ -221,13 +221,12 @@ bool ReadBufferFromS3::nextImpl()
 
 size_t ReadBufferFromS3::readBigAt(char * to, size_t n, size_t range_begin, const std::function<bool(size_t)> & progress_callback)
 {
-    if (n == 0)
-        return 0;
-
+    size_t initial_n = n;
     size_t sleep_time_with_backoff_milliseconds = 100;
-    for (size_t attempt = 0;; ++attempt)
+    for (size_t attempt = 0; n > 0; ++attempt)
     {
         bool last_attempt = attempt + 1 >= request_settings.max_single_read_retries;
+        size_t bytes_copied = 0;
 
         ProfileEventTimeIncrement<Microseconds> watch(ProfileEvents::ReadBufferFromS3Microseconds);
 
@@ -236,14 +235,12 @@ size_t ReadBufferFromS3::readBigAt(char * to, size_t n, size_t range_begin, cons
             auto result = sendRequest(range_begin, range_begin + n - 1);
             std::istream & istr = result.GetBody();
 
-            size_t bytes = copyFromIStreamWithProgressCallback(istr, to, n, progress_callback);
+            copyFromIStreamWithProgressCallback(istr, to, n, progress_callback, &bytes_copied);
 
-            ProfileEvents::increment(ProfileEvents::ReadBufferFromS3Bytes, bytes);
+            ProfileEvents::increment(ProfileEvents::ReadBufferFromS3Bytes, bytes_copied);
 
             if (read_settings.remote_throttler)
-                read_settings.remote_throttler->add(bytes, ProfileEvents::RemoteReadThrottlerBytes, ProfileEvents::RemoteReadThrottlerSleepMicroseconds);
-
-            return bytes;
+                read_settings.remote_throttler->add(bytes_copied, ProfileEvents::RemoteReadThrottlerBytes, ProfileEvents::RemoteReadThrottlerSleepMicroseconds);
         }
         catch (Poco::Exception & e)
         {
@@ -253,7 +250,13 @@ size_t ReadBufferFromS3::readBigAt(char * to, size_t n, size_t range_begin, cons
             sleepForMilliseconds(sleep_time_with_backoff_milliseconds);
             sleep_time_with_backoff_milliseconds *= 2;
         }
+
+        range_begin += bytes_copied;
+        to += bytes_copied;
+        n -= bytes_copied;
     }
+
+    return initial_n;
 }
 
 bool ReadBufferFromS3::processException(Poco::Exception & e, size_t read_offset, size_t attempt) const

--- a/src/IO/ReadWriteBufferFromHTTP.cpp
+++ b/src/IO/ReadWriteBufferFromHTTP.cpp
@@ -594,16 +594,14 @@ size_t ReadWriteBufferFromHTTPBase<UpdatableSessionPtr>::readBigAt(char * to, si
     /// This ensures we've sent at least one HTTP request and populated saved_uri_redirect.
     chassert(file_info && file_info->seekable);
 
-    if (n == 0)
-        return 0;
-
     Poco::URI uri_ = saved_uri_redirect.value_or(uri);
     if (uri_.getPath().empty())
         uri_.setPath("/");
 
+    size_t initial_n = n;
     size_t milliseconds_to_wait = settings.http_retry_initial_backoff_ms;
 
-    for (size_t attempt = 0;; ++attempt)
+    for (size_t attempt = 0; n > 0; ++attempt)
     {
         bool last_attempt = attempt + 1 >= settings.http_max_tries;
 
@@ -616,6 +614,7 @@ size_t ReadWriteBufferFromHTTPBase<UpdatableSessionPtr>::readBigAt(char * to, si
 
         Poco::Net::HTTPResponse response;
         std::istream * result_istr;
+        size_t bytes_copied = 0;
 
         try
         {
@@ -629,17 +628,14 @@ size_t ReadWriteBufferFromHTTPBase<UpdatableSessionPtr>::readBigAt(char * to, si
                     "Expected 206 Partial Content, got {} when reading {} range [{}, {})",
                     toString(response.getStatus()), uri_.toString(), offset, offset + n);
 
-            bool cancelled;
-            size_t r = copyFromIStreamWithProgressCallback(*result_istr, to, n, progress_callback, &cancelled);
-
-            if (!cancelled)
+            copyFromIStreamWithProgressCallback(*result_istr, to, n, progress_callback, &bytes_copied);
+            if (bytes_copied == n)
             {
+                result_istr->ignore(UINT64_MAX);
                 /// Response was fully read.
-                markSessionForReuse(sess);
+                markSessionForReuse(*sess);
                 ProfileEvents::increment(ProfileEvents::ReadWriteBufferFromHTTPPreservedSessions);
             }
-
-            return r;
         }
         catch (const Poco::Exception & e)
         {
@@ -664,9 +660,15 @@ size_t ReadWriteBufferFromHTTPBase<UpdatableSessionPtr>::readBigAt(char * to, si
 
             sleepForMilliseconds(milliseconds_to_wait);
             milliseconds_to_wait = std::min(milliseconds_to_wait * 2, settings.http_retry_max_backoff_ms);
-            continue;
         }
+
+        /// Make sure retries don't re-read the bytes that we've already reported to progress_callback.
+        offset += bytes_copied;
+        to += bytes_copied;
+        n -= bytes_copied;
     }
+
+    return initial_n;
 }
 
 template <typename UpdatableSessionPtr>

--- a/src/IO/SeekableReadBuffer.cpp
+++ b/src/IO/SeekableReadBuffer.cpp
@@ -64,7 +64,7 @@ std::unique_ptr<SeekableReadBuffer> wrapSeekableReadBufferPointer(SeekableReadBu
     return std::make_unique<SeekableReadBufferWrapper<SeekableReadBufferPtr>>(*ptr, SeekableReadBufferPtr{ptr});
 }
 
-size_t copyFromIStreamWithProgressCallback(std::istream & istr, char * to, size_t n, const std::function<bool(size_t)> & progress_callback, bool * out_cancelled)
+void copyFromIStreamWithProgressCallback(std::istream & istr, char * to, size_t n, const std::function<bool(size_t)> & progress_callback, size_t * out_bytes_copied, bool * out_cancelled)
 {
     const size_t chunk = DBMS_DEFAULT_BUFFER_SIZE;
     if (out_cancelled)
@@ -82,6 +82,7 @@ size_t copyFromIStreamWithProgressCallback(std::istream & istr, char * to, size_
         bool cancelled = false;
         if (gcount && progress_callback)
             cancelled = progress_callback(copied);
+        *out_bytes_copied = copied;
 
         if (gcount != to_copy)
         {
@@ -103,7 +104,7 @@ size_t copyFromIStreamWithProgressCallback(std::istream & istr, char * to, size_
         }
     }
 
-    return copied;
+    *out_bytes_copied = copied;
 }
 
 }

--- a/src/IO/SeekableReadBuffer.h
+++ b/src/IO/SeekableReadBuffer.h
@@ -98,6 +98,7 @@ std::unique_ptr<SeekableReadBuffer> wrapSeekableReadBufferReference(SeekableRead
 std::unique_ptr<SeekableReadBuffer> wrapSeekableReadBufferPointer(SeekableReadBufferPtr ptr);
 
 /// Helper for implementing readBigAt().
-size_t copyFromIStreamWithProgressCallback(std::istream & istr, char * to, size_t n, const std::function<bool(size_t)> & progress_callback, bool * out_cancelled = nullptr);
+/// Updates *out_bytes_copied after each call to the callback, as well as at the end.
+void copyFromIStreamWithProgressCallback(std::istream & istr, char * to, size_t n, const std::function<bool(size_t)> & progress_callback, size_t * out_bytes_copied, bool * out_cancelled = nullptr);
 
 }

--- a/tests/integration/test_storage_s3/test.py
+++ b/tests/integration/test_storage_s3/test.py
@@ -1018,11 +1018,11 @@ def test_url_reconnect_in_the_middle(started_cluster):
         def select():
             global result
             result = instance.query(
-                f"""select sum(cityHash64(x)) from (select toUInt64(id) + sleep(0.1) as x from
+                f"""select count(), sum(cityHash64(x)) from (select toUInt64(id) + sleep(0.1) as x from
                 url('http://{started_cluster.minio_host}:{started_cluster.minio_port}/{bucket}/{filename}', 'TSV', '{table_format}')
                 settings http_max_tries = 10, http_retry_max_backoff_ms=2000, http_send_timeout=1, http_receive_timeout=1)"""
             )
-            assert int(result) == 3914219105369203805
+            assert result == "1000000\t3914219105369203805\n"
 
         thread = threading.Thread(target=select)
         thread.start()
@@ -1035,7 +1035,7 @@ def test_url_reconnect_in_the_middle(started_cluster):
 
         thread.join()
 
-        assert int(result) == 3914219105369203805
+        assert result == "1000000\t3914219105369203805\n"
 
 
 def test_seekable_formats(started_cluster):

